// mjreport/dist/Common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Tags;
(function (Tags) {
    Tags[Tags["Text"] = 0] = "Text";
    Tags[Tags["Heading"] = 1] = "Heading";
    Tags[Tags["Table"] = 2] = "Table";
    Tags[Tags["PageBreak"] = 3] = "PageBreak";
})(Tags = exports.Tags || (exports.Tags = {}));
var HeadingLevel;
(function (HeadingLevel) {
    HeadingLevel["H1"] = "h1";
    HeadingLevel["H2"] = "h2";
    HeadingLevel["H3"] = "h3";
    HeadingLevel["H4"] = "h4";
    HeadingLevel["H5"] = "h5";
    HeadingLevel["H6"] = "h6";
})(HeadingLevel = exports.HeadingLevel || (exports.HeadingLevel = {}));
var TableReservedField;
(function (TableReservedField) {
    TableReservedField["index"] = "index";
    TableReservedField["row"] = "row";
    TableReservedField["empty"] = "empty";
})(TableReservedField = exports.TableReservedField || (exports.TableReservedField = {}));
var TableCellType;
(function (TableCellType) {
    TableCellType["Header"] = "th";
    TableCellType["Body"] = "td";
})(TableCellType = exports.TableCellType || (exports.TableCellType = {}));
function isTableCell(obj) {
    return Array.isArray(obj) && typeof obj[0] === 'object';
}
exports.isTableCell = isTableCell;
function isTableField(obj) {
    return !Array.isArray(obj) && typeof obj === 'function';
}
exports.isTableField = isTableField;
exports.DefaultStyles = "\nbody {\n  background: rgb(204,204,204); \n}\n\npage {\n  background: white;\n  margin: 0 auto;\n  margin-bottom: 0.5cm;\n  box-shadow: 0 0 0.5cm rgba(0,0,0,0.5);\n  display: flex;\n  flex-direction: column;\n}\n\npage[type=\"A4_Portrait\"] {\n  width: 210mm;\n  height: 297mm;\n}\n\npage[type=\"A4_Landscape\"] {\n  width: 297mm;\n  height: 210mm;\n}\n\npage[type=\"A5_Portrait\"] {\n  width: 148mm;\n  height: 210mm;\n}\n\npage[type=\"A5_Landscape\"] {\n  width: 210mm;\n  height: 148mm;\n}\n\n@media print {\n  .no-print {\n    display: none;\n  } \n  \n  body, page {\n    margin: 0;\n    box-shadow: none;\n  }\n}\n\n.header_section {dir:rtl;\n  margin-top: 10px;\n  margin-left: 5px;\n  margin-right: 5px;\n  flex-shrink: 0;\n  position: relative;\n}\n\n.content_section {\n  margin-left: 5px;\n  margin-right: 5px;\n  height: 100%;\n  overflow: hidden;\n}\n\n.footer_section {dir:rtl;\n  margin-bottom: 5px;\n  margin-left: 5px;\n  margin-right: 5px;\n  flex-shrink: 0;\n  overflow: hidden;\n  position: relative;\n}\n\n.basic_tbl {\n  border: 1px solid black;\n  border-collapse: collapse;\n  margin: auto;\n  padding:5px;\n  width: 100%;\n}\n\n.basic_tbl th {\n  border-bottom: 2px solid black;\n  border-right: 1px solid black;\n  font-size: 20px;\n  font-weight: 600;\n  padding: 5px;\n  vertical-align: middle;\n}\n\n.basic_tbl tr {\n  border-top: 1px solid black;\n  border-bottom: 1px solid black;\n}\n\n.basic_tbl tr:hover td {\n  background:rgb(219, 222, 255);\n}\n\n.basic_tbl tr:nth-child(odd) td {\n  background:#EBEBEB;\n}\n\n.basic_tbl tr:nth-child(odd):hover td {\n  background:rgb(219, 222, 255);\n}\n\n.basic_tbl td {min-width:100px;\n  font-size:11pt;\n  vertical-align:middle;\n  padding:5px;\n  border-right: 1px solid black;\n  border-left: 1px solid black;\n} \n\n.rtl {\n  direction: rtl;\n}\n\n.ltr {\n  direction: ltr;\n}\n\n.text-left {\n  text-align: left;\n}\n\n.text-center {\n  text-align: center;\n}\n\n.text-right {\n  text-align: right;\n}\n";
